### PR TITLE
feat(dynamodb): deprecate $AWS.DynamoDB, add methods to Table instance

### PR DESCRIPTION
**Pull request by sam-goodwin**
This change adds some common operations to the `Table`. These types include some of the type safety we were enjoying from typesafe-dynamodb except the expensive features such as ProjectionExpression and type-safe ExpressionAttributeNames/Values are removed.
- [x] getItem 
- [x] putItem
- [x] deleteItem
- [x] updateITem
- [x] batchGetItems
- [x] batchWriteItems
- [x] transactGet
- [x] transactWrite
- [x] appsync apis
- [x] update website/docs
- [x] update code docs

This change also introduces a new approach to better namespace variants of APIs. For example, let's consider `table.get`

```ts
// we prioritize for lambda on this interface - so it uses the DocumentClient
table.get({
  key: "key"
});

// `.attributes` represents the low-level DynamoDB AttributeValue format, it can be used in Step Functions
table.attributes.get({
  key: {
    S: "key",
  }
);

// appsync's one-off interface
table.appsync.get({
  key: {
    S: "key",
  },
  expressionValues: {
    ..
  }
});
```

This namespace approach will scale quite well for all services. Tenets:
1. always prioritize Lambda for the root interface. If it can support step functions/appsync, then great, but not mandatory
2. add named method properties on the interface for variants, for example `appsync` or `attributes` for dynamodb.

### CODE REVIEW

File: src/table/appsync.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      ExpressionAttributeNames: "expressionNames";
+    }
+  >;
+
+type RenameKeys<
+  T extends object,
+  Substitutions extends Record<string, string>
+> = {
+  [k in keyof T as k extends keyof Substitutions ? Substitutions[k] : k]: T[k];
+};
```
Comment (sam-goodwin) [inline src/table/appsync.ts:75]: Need to remove

File: src/table/batch-get-item.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+import { NodeKind } from "../node-kind";
+import { addIfDefined, makeAppSyncTableIntegration } from "./appsync";
+import {
+  createDynamoAttributesIntegration,
+  createDynamoDocumentIntegration,
+} from "./integration";
+import { ITable } from "./table";
+import { AttributeKeyToObject } from "./util";
+
+export interface GetItemInput<
```
Comment (sam-goodwin) [inline src/table/batch-get-item.ts:15]: Wrong name

File: test/__snapshots__/step-function.test.ts.snap [hunk 3]
```diff
@@ -6401,17 +6401,15 @@ exports[`call AWS.DynamoDB.GetItem, then Lambda and return LiteralExpr 1`] = `
...
       "Parameters": {
-        "Key": {
-          "id": {
-            "S.$": "$.input.id",
-          },
-        },
         "TableName": "__REPLACED_TOKEN",
+        "id": {
+          "S.$": "$.input.id",
+        },
```
Comment (thantos) [inline test/__snapshots__/step-function.test.ts.snap:6410]: regression?

File: test/__snapshots__/validate.test.ts.snap [hunk 4]
```diff
@@ -13,6 +34,12 @@ https://functionless.org/docs/error-codes#awsmethod-request-must-have-exactly-on
 [7m  [0m [91m~~~~~~~~~~~~~~~~~~~~~~~[0m
 [7m77[0m   },
 [7m  [0m [91m~~~[0m
+[96mtest/test-files/api-gateway.ts[0m:[93m88[0m:[93m11[0m - [91merror[0m[90m Functionless(10029): [0mApiGateway Unsupported Reference: $AWS
+
+https://functionless.org/docs/error-codes#apigateway-unsupported-reference
+
+[7m88[0m     return $AWS.DynamoDB.GetItem({
+[7m  [0m [91m          ~~~~~[0m
 [96mtest/test-files/api-gateway.ts[0m:[93m89[0m:[93m20[0m - [91merror[0m[90m Functionless(10011): [0mAPI Gateway does not support spread assignment expressions
```
Comment (thantos) [inline test/__snapshots__/validate.test.ts.snap:43]: this looks wrong?

**Review by thantos — CHANGES_REQUESTED** [review]

File: test/__snapshots__/step-function.test.ts.snap (hunk 3 above)
Comment (sam-goodwin) [reply test/__snapshots__/step-function.test.ts.snap:6410]: Yeah I've changed the interface but haven't updated the implementation. Will do that today

File: src/table/integration.ts [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+      }
+      grantPermissions(context.role);
+
+      const input = call.args[0]?.expr;
+      if (input === undefined) {
+        return context.stateWithHeapOutput({
+          Type: "Task",
+          Resource: `arn:aws:states:::aws-sdk:dynamodb:${operationName}`,
+          Parameters: {
+            TableName: table.tableName,
```
[the remark this answers is not in the document]
Comment (sam-goodwin) [reply src/table/integration.ts:114]: I thought I ran into an error when using name instead of arn. Let me try again

File: src/table/integration.ts [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+          },
+          Next: ASLGraph.DeferNext,
+        });
+      } else {
+        return context.evalExpr(input, (output) => {
+          return context.stateWithHeapOutput({
+            Type: "Task",
+            Resource: `arn:aws:states:::aws-sdk:dynamodb:${operationName}`,
+            Parameters: {
+              ...output.value,
```
[the remark this answers is not in the document]
Comment (sam-goodwin) [reply src/table/integration.ts:135]: It was `any` for some reason - I have updated the Step Functions code and tested them with real step functions. Working nicely now. There was indeed a problem here.

File: src/table/transact-get-item.ts [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+                  states: {
+                    "set TableName": {
+                      Type: "Pass",
+                      Result: tableName,
+                      ResultPath: `$.item.Get.TableName`,
+                      OutputPath: "$.item",
+                      End: true,
+                    },
+                  },
+                }),
```
Comment (thantos) [inline src/table/transact-get-item.ts:137]: nit: you can pass in just a state an "override name" into `aslGraphToStates`.

File: src/table/transact-get-item.ts [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+            },
+            startState: "set TableName",
+            states: {
+              "set TableName": {
+                Type: "Map",
+                ResultPath: `${output.jsonPath}.TransactItems`,
+                Parameters: {
+                  "item.$": "$$.Map.Item.Value",
+                },
+                ItemsPath: `${output.jsonPath}.TransactItems`,
```
Comment (thantos) [inline src/table/transact-get-item.ts:125]: FYI: you can do this now `ASLGraph.jsonPath(output, "TransactItems").jsonPath`

File: src/table/transact-get-item.ts [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+    },
+    {
+      override: (input, context) => {
+        return context.evalExprToJsonPath(input, (output) => {
+          const heapVar = context.newHeapVariable();
+
+          return <ASLGraph.OutputSubState>{
+            output: {
+              jsonPath: heapVar,
+            },
```
Comment (thantos) [inline src/table/transact-get-item.ts:116]: nit: `ASLGraph.jsonPath(heapVar)`

File: src/table/transact-write-item.ts [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+                    } as any;
+
+                    function branch(
+                      name: keyof AWS.DynamoDB.TransactWriteItem
+                    ) {
+                      return {
+                        Variable: `$.item.${name}`,
+                        IsPresent: true,
+                        Next: `set TableName on ${name}`,
+                      };
```
Comment (thantos) [inline src/table/transact-write-item.ts:207]: nit: `{ ...ASL.isPresent(...), Next: "" }`

File: src/table/transact-write-item.ts [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+                        ],
+                        Default: "fail",
+                      },
+                      ...set("ConditionCheck"),
+                      ...set("Delete"),
+                      ...set("Put"),
+                      ...set("Update"),
+                      ["fail"]: {
+                        Type: "Fail",
+                        Error: "bad",
```
Comment (thantos) [inline src/table/transact-write-item.ts:196]: bad

Comment (thantos) [reply src/table/transact-write-item.ts:196]: I do have a bit of a pattern for runtime errors

```ts
{
                      Type: "Fail",
                      Error: "Functionless.InvalidAccess",
                      Cause:
                        "Reference element access is not valid for objects.",
                    }
```

File: test/table.test.ts [hunk 12]
```diff
@@ -76,7 +76,7 @@ export async function typeCheck() {
   t2 = t3;
 
   t3 = t1;
-  t3 = t2;
+  // t3 = t2; // regression: https://github.com/functionless/functionless/pull/519
```
Comment (thantos) [inline test/table.test.ts:79]: ?

**Review by thantos — APPROVED** [review]

**Review by thantos — APPROVED** [review]

File: test/table.test.ts (hunk 12 above)
Comment (sam-goodwin) [reply test/table.test.ts:79]: `Table<Item>` is no longer a sub-type of `Table<Record<string, any>>`.

Would love to fix it but it is not a priority IMO.
